### DISCUSSION TITLE
how do I modify the resolution?

### DISCUSSION THREAD

**Discussion by jjplano** [Roger-random/ESP_8_BIT_composite#36] [Q&A] [answered]
hi! fantastic work, congrats on this! always really appreciated by the experimental analog video community to be able to add some tools

I wanted to ask how to change the resolution of the sketches, if that's possible. I've noted that on NTSC the image reaches the edges of the screen, but on PAL that doesn't happen.

thanks!

**Comment by Roger-random** [comment, marked as answer]
Thank you for your appreciation!

Unfortunately, it is quite difficult to change display resolution. The original [ESP_8_BIT](https://github.com/rossumur/esp_8_bit) project (from which this library was derived) used a lot of clever code to generate TV analog signal including precomputed lookup tables. (Example: [phase representation for color palette.](https://github.com/Roger-random/ESP_8_BIT_composite/blob/94bde82651c4c9095d6de369cc44b04265302e7a/ESP_8_BIT_composite.cpp#L178)) I designed this library to be a beginner-friendly Arduino library, and the knowledge necessary to recalculate data tables are not beginner-friendly.

Advanced users who are willing to dive into such work can look at the ESP_8_BIT project and learn from the differences between its Atari, SEGA, and Nintendo emulation modes. Each of those modes have a different resolution and color palette, serving as examples for custom resolutions. (My library here used the SEGA mode.) 

Another option to examine is [FabGL](http://www.fabglib.org/), which is a more powerful (but also more complex) graphics library that supports multiple resolutions. When I started my project, FabGL did not support composite analog video. But that capability has since been added to FabGL. (See the author's [show-and-tell thread.](https://github.com/fdivitto/FabGL/discussions/192)) An additional reason to try FabGL is that the author developed it with an actual composite signal analysis instrument (Tektronix VM700) for PAL video, so I would expect its PAL signal to be rock solid. (I lack such equipment, so my analysis is the much less reliable "Looks good on my TV".)

**Comment by jjplano** [comment]
thank you for your answer! yeah, I want to explore the possibility of building a cheap diy visual tool based on this board, as what I'm seeing the possibilities are really good. but I'm just on the exploratory phase for now. so getting the correct resolutions from the start is something to keep in mind, so I don't get too deep into something that can't be changed later. gonna check FabGL, sounds really good. thanks again!
